File: sdks/rs/src/clients/obligations/erc1155/payment.rs
//! ERC1155 payment obligation client
//!
//! Provides functionality for making direct ERC1155 token payments.

use alloy::primitives::{Address, Bytes, FixedBytes};
use alloy::rpc::types::TransactionReceipt;
use alloy::sol_types::SolValue;

use crate::contracts;
use crate::types::{DecodedAttestation, Erc1155Data};

use super::Erc1155Module;

/// Payment API for ERC1155 tokens
pub struct Payment<'a> {
    module: &'a Erc1155Module,
}

impl<'a> Payment<'a> {
    pub fn new(module: &'a Erc1155Module) -> Self {
        Self { module }
    }

    /// Get the contract address
    pub fn address(&self) -> Address {
        self.module.addresses.payment_obligation
    }

    /// Decodes ERC1155PaymentObligation.ObligationData from bytes.
    ///
    /// # Arguments
    /// * `obligation_data` - The obligation data
    ///
    /// # Returns
    /// * `Result<contracts::obligations::ERC1155PaymentObligation::ObligationData>` - The decoded obligation data
    pub fn decode_obligation(
        obligation_data: &Bytes,
    ) -> eyre::Result<contracts::obligations::ERC1155PaymentObligation::ObligationData> {
        let obligation_data =
            contracts::obligations::ERC1155PaymentObligation::ObligationData::abi_decode(
                obligation_data,
            )?;
        Ok(obligation_data)
    }

    /// Gets a payment obligation by its attestation UID.
    pub async fn get_obligation(
        &self,
        uid: FixedBytes<32>,
    ) -> eyre::Result<DecodedAttestation<contracts::obligations::ERC1155PaymentObligation::ObligationData>>
    {
        let eas_contract =
            contracts::IEAS::new(self.module.addresses.eas, &self.module.wallet_provider);

        let attestation = eas_contract.getAttestation(uid).call().await?;
        let obligation_data =
            contracts::obligations::ERC1155PaymentObligation::ObligationData::abi_decode(
                &attestation.data,
            )?;

        Ok(DecodedAttestation {
            attestation,
            data: obligation_data,
        })
    }

    /// Makes a direct payment with ERC1155 tokens.
    ///
    /// # Arguments
    /// * `price` - The ERC1155 token data for payment
    /// * `payee` - The address of the payment recipient
    ///
    /// # Returns
    /// * `Result<TransactionReceipt>` - The transaction receipt
    pub async fn pay(
        &self,
        price: &Erc1155Data,
        payee: Address,
    ) -> eyre::Result<TransactionReceipt> {
        let payment_obligation_contract = contracts::obligations::ERC1155PaymentObligation::new(
            self.module.addresses.payment_obligation,
            &self.module.wallet_provider,
        );

        let receipt = payment_obligation_contract
            .doObligation(
                contracts::obligations::ERC1155PaymentObligation::ObligationData {
                    token: price.address,
                    tokenId: price.id,
                    amount: price.value,
                    payee,
                },
                FixedBytes::<32>::ZERO, // refUID - no reference for standalone payments
            )
            .send()
            .await?
            .get_receipt()
            .await?;

        Ok(receipt)
    }
}

File: sdks/ts/src/clients/obligations/erc1155/payment.ts
import { decodeAbiParameters, encodeAbiParameters, getAbiItem } from "viem";
import { abi as erc1155PaymentAbi } from "../../../contracts/obligations/payment/ERC1155PaymentObligation";
import type { Erc1155 } from "../../../types";
import { getAttestation, getAttestedEventFromTxHash, writeContract, type ViemClient } from "../../../utils";
import type { Erc1155Addresses } from "./index";
import { makeErc1155UtilClient } from "./util";

const erc1155PaymentDecodeFunction = getAbiItem({
  abi: erc1155PaymentAbi.abi,
  name: "decodeObligationData",
});

const erc1155PaymentObligationDataType = erc1155PaymentDecodeFunction.outputs[0];

export type Erc1155PaymentClient = ReturnType<typeof makeErc1155PaymentClient>;

export const makeErc1155PaymentClient = (
  viemClient: ViemClient,
  addresses: Erc1155Addresses,
) => {
  const util = makeErc1155UtilClient(viemClient, addresses);

  const getSchema = async () =>
    await viemClient.readContract({
      address: addresses.paymentObligation,
      abi: erc1155PaymentAbi.abi,
      functionName: "ATTESTATION_SCHEMA",
      authorizationList: undefined,
    });

  return {
    address: addresses.paymentObligation,
    getSchema,

    encodeObligationRaw: (data: { token: `0x${string}`; tokenId: bigint; amount: bigint; payee: `0x${string}` }) => {
      return encodeAbiParameters([erc1155PaymentObligationDataType], [data]);
    },

    encodeObligation: (token: Erc1155, payee: `0x${string}`) => {
      return encodeAbiParameters([erc1155PaymentObligationDataType], [{
        token: token.address,
        tokenId: token.id,
        amount: token.value,
        payee,
      }]);
    },

    decodeObligation: (obligationData: `0x${string}`) => {
      return decodeAbiParameters([erc1155PaymentObligationDataType], obligationData)[0];
    },

    getObligation: async (uid: `0x${string}`) => {
      const [attestation, schema] = await Promise.all([getAttestation(viemClient, uid), getSchema()]);

      if (attestation.schema !== schema) {
        throw new Error(`Unsupported schema: ${attestation.schema}`);
      }
      const data = decodeAbiParameters([erc1155PaymentObligationDataType], attestation.data)[0];

      return {
        ...attestation,
        data,
      };
    },

    pay: async (price: Erc1155, payee: `0x${string}`, refUID: `0x${string}` = "0x0000000000000000000000000000000000000000000000000000000000000000") => {
      const hash = await writeContract(viemClient, {
        address: addresses.paymentObligation,
        abi: erc1155PaymentAbi.abi,
        functionName: "doObligation",
        args: [
          {
            token: price.address,
            tokenId: price.id,
            amount: price.value,
            payee,
          },
          refUID,
        ],
      });

      const attested = await getAttestedEventFromTxHash(viemClient, hash);
      return { hash, attested };
    },

    approveAndPay: async (price: Erc1155, payee: `0x${string}`, refUID: `0x${string}` = "0x0000000000000000000000000000000000000000000000000000000000000000") => {
      await util.approveAll(price.address, "payment");
      const hash = await writeContract(viemClient, {
        address: addresses.paymentObligation,
        abi: erc1155PaymentAbi.abi,
        functionName: "doObligation",
        args: [
          {
            token: price.address,
            tokenId: price.id,
            amount: price.value,
            payee,
          },
          refUID,
        ],
      });

      const attested = await getAttestedEventFromTxHash(viemClient, hash);
      return { hash, attested };
    },
  };
};
